### PR TITLE
[Internal][Executor] Refine the LineExecutionProcessPool to align more closely with a standard process pool

### PR DESCRIPTION
**Pull request by PeiwenGaoMS**
# Description

## Background
After splitting the executor and runtime into two separate containers, the implementation of batch run in the executor server also requires a process pool to handle execution requests line by line. Therefore, we need the line process pool to have the capability to process individual lines. As a process pool, it should also have the ability to submit one or multiple tasks at a time, consistent with the python process pool interface.

## Usages
- Use with context
``` python
with LineExecutionProcessPool(...) as pool:
      line_results = await pool.run(zip(line_number, batch_inputs))
```
- Use method
``` python
pool = LineExecutionProcessPool(...)
pool.start()
line_results = await pool.run(zip(line_number, batch_inputs))
pool.close()
```

## Public Functions
- `start`: Create task queue, input/output queues, and start processes and monitor thread pool.
- `close`: Send terminate signal to monitor threads, end processes and close thread pool.
- `run`: Put all line inputs to task queue and get the line results list.
- `submit`: Put one line input to task queue and get one line result.

## Implementation Process
image

## Main Differences from Previous Implementation
- Start the monitor threads at the beginning instead of doing it in the `run` method.
- The monitor thread will not exit due to an empty task queue. It exits only when the batch run times out or when a termination signal is received from the task queue.
- Update `run` to async function.

## Code Changes Sunmary
This pull request includes changes to various parts of the `promptflow` system, with the main focus being the addition of new functionalities and the refactoring of existing code for better performance and readability. The most significant changes include the addition of a new function to convert multimedia data to string, the introduction of a method to determine the maximum number of workers that can be created, the conversion of some functions to asynchronous, and the implementation of new exception classes.

New functionalities:

* [`src/promptflow/promptflow/_utils/multimedia_utils.py`](diffhunk://#diff-46760b7ed265c4b29c9cc4111fb19da9b815717e4d7180f61e0d0da023d2a8f9R181-R185): Added a new function `convert_multimedia_data_to_string` to convert multimedia data to string.
* [`src/promptflow/promptflow/_utils/process_utils.py`](diffhunk://#diff-25b180c7dfe70758ae52d59d4572a4bc24fe331b916575c8286ab5af05f112e4R5-R64): Introduced a new function `get_available_max_worker_count` to determine the maximum number of workers that can be created.
* [`src/promptflow/promptflow/executor/_errors.py`](diffhunk://#diff-1f02407850413baccbb33e0078d9772e15ae1c46c020dc7457529c6f7a054315R192-R197): Implemented new exception classes `ThreadCrashError` and `ProcessCrashError` to handle thread and process crashes.

Refactoring:

* `src/promptflow/promptflow/batch/_batch_engine.py`, `src/promptflow/promptflow/batch/_python_executor_proxy.py`: Converted some functions to asynchronous for better performance. [[1]](diffhunk://#diff-ecf0905f2116abe08b5cf2931b856bf39aa8b38a30fadd9042538d076dbfde80L349-R349) [[2]](diffhunk://#diff-8c42522c7186705de9a6651343af6006be02b7342f72467c2785f23f93ea669eL47-R47) [[3]](diffhunk://#diff-8c42522c7186705de9a6651343af6006be02b7342f72467c2785f23f93ea669eL82-R82)
* [`src/promptflow/promptflow/executor/_process_manager.py`](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R4-R23): Refactored the `AbstractProcessManager` and `SpawnProcessManager` classes, added new methods for process management, and made changes to improve readability. [[1]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R4-R23) [[2]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R58-R60) [[3]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L61-R72) [[4]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R113-R141) [[5]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L177-R217) [[6]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L337-R374) [[7]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L369-R406) [[8]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L437-R471) [[9]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258L446-R480) [[10]](diffhunk://#diff-fb203fa9d75078068758d8934fd7711b63934b15faa6cee7c3c4d16ab3347258R529-R549)
* `src/promptflow/tests/executor/e2etests/test_batch_timeout.py`, `src/promptflow/tests/executor/unittests/_utils/test_process_utils.py`, `src/promptflow/tests/executor/unittests/executor/test_line_execution_process_pool.py`: Updated test cases to reflect the changes made in the codebase. [[1]](diffhunk://#diff-e79471098e1447cae89a66ed2b5e2194ca7dfa2317d994f261de3eaf1be63c16R1-R24) [[2]](diffhunk://#diff-7888cd3b10ffbfe09c0035975a082a8e36481a9aa57818da5c5e7c2a6db8dea4R1-R38) [[3]](diffhunk://#diff-8b545092ba1c4e2536841f46825c30b22ba56195407de5b115ee5ca583c67fa6L22) [[4]](diffhunk://#diff-8b545092ba1c4e2536841f46825c30b22ba56195407de5b115ee5ca583c67fa6R186-R193)

Additions:

* [`src/promptflow/promptflow/storage/_queue_run_storage.py`](diffhunk://#diff-e79471098e1447cae89a66ed2b5e2194ca7dfa2317d994f261de3eaf1be63c16R1-R24): Added a new storage class `QueueRunStorage` for storing run information in a queue.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 1]
```diff
@@ -185,121 +166,187 @@ def __enter__(self):
...
+        exit_loop = False
+        while not exit_loop:
+            while True:
+                try:
+                    # Get task from task_queue
+                    data = task_queue.get(timeout=1)
+                    # Calculate the line timeout for the current line.
+                    line_timeout_sec = self._calculate_line_timeout_sec()
+                    # If the task is a terminate signal or the batch run is timeouted, exit the loop.
+                    if data == TERMINATE_SIGNAL or line_timeout_sec is None:
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:305]: Why line_timeout_sec is None means batch timeout?

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:305]: image
This function will return None if remaining_execution_time <= 0.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 2]
```diff
@@ -185,121 +166,187 @@ def __enter__(self):
...
-                    self._is_timeout = True
+        # The main loop of the thread, responsible for getting tasks from the task queue and
+        # processing them through the input queue, while also monitoring for terminate signals.
+        # Currently, it exits this loop only upon receiving a terminate signal (TERMINATE_SIGNAL).
+        exit_loop = False
+        while not exit_loop:
+            while True:
+                try:
+                    # Get task from task_queue
+                    data = task_queue.get(timeout=1)
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:301]: It looks like a sync wait, not an async wait?

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:301]: The `_monitor_workers_and_process_tasks_in_thread` is a sync function.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 3]
```diff
@@ -112,20 +76,26 @@ def __init__(
...
             bulk_logger.info(f"Set start method to default {multiprocessing.get_start_method()}.")
             multiprocessing_start_method = multiprocessing.get_start_method()
-        use_fork = multiprocessing_start_method in ["fork", "forkserver"]
-        self._flow_file = flow_executor._flow_file
-        self._connections = flow_executor._connections
-        self._working_dir = flow_executor._working_dir
-        self._use_fork = use_fork
+        self._use_fork = multiprocessing_start_method in ["fork", "forkserver"]
+
+        # Initialize some fields from inputs.
```
Comment (Hhhilulu) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:81]: Maybe _batch_timeout_sec/_line_timeout_sec  is not from inputs?

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:81]: The inputs here refers to the inputs of the init method. I will update it to make clear.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 4]
```diff
@@ -185,121 +166,188 @@ def __enter__(self):
...
-
-        return self
-
-    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Start a thread pool to monitor the processes.
+        self._monitor_pool = ThreadPool(
+            self._n_process, initializer=set_context, initargs=(contextvars.copy_context(),)
+        )
+        # The variable 'async_result' here is not the actual result of the batch run
+        # but an AsyncResult object that can be used to check if the execution are finished
```
[the remark this answers is not in the document]
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:174]: Updated.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 5]
```diff
@@ -185,121 +166,188 @@ def __enter__(self):
...
-        while not self._batch_timeout_expired(batch_start_time):
-            self._processes_manager.ensure_healthy()
-            try:
-                # Get task from task_queue
-                inputs, line_number, run_id = task_queue.get(timeout=1)
-            except queue.Empty:
-                break
+        # Get the process info of the thread monitoring from the manager.
+        index, process_id, process_name = self._processes_manager.get_process_info(index)
+        self._processes_manager.ensure_healthy()
```
Comment (Hhhilulu) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:292]: Move it[self._processes_manager.ensure_healthy()] to the while loop below.

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Updated.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 6]
```diff
@@ -185,121 +166,188 @@ def __enter__(self):
...
-        return psutil.pid_exists(process_id)
+        if self._batch_timeout_expired(self._batch_start_time):
+            # Send terminate signal to all threads and wait for them to exit.
+            self._terminate_tasks()
+            # Wait for up to 10s for thread termination, aiming to ensure that
+            # the line results in the result dict are as complete as possible.
+            start_time = datetime.utcnow()
+            while not self._timeout_expired(start_time, self._THREAD_TERMINATED_TIMEOUT):
+                if self._all_tasks_ready():
+                    break
```
Comment (Hhhilulu) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:274]: Better to add time.sleep(1) to avoid memory brust.

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:274]: Updated.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 7]
```diff
@@ -185,121 +166,189 @@ def __enter__(self):
...
+                bulk_logger.error(f"ProcessPool failed with exception: {e}")
+                raise ProcessPoolError(
+                    message_format=f"ProcessPool failed with exception: {e}",
+                    target=ErrorTarget.EXECUTOR,
+                ) from e
 
-    def _is_process_alive(self, process_id):
-        return psutil.pid_exists(process_id)
+        if self._batch_timeout_expired(self._batch_start_time):
+            # Send terminate signal to all threads and wait for them to exit.
```
Comment (Hhhilulu) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:267]: Send terminate signal to process not thread?

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 8]
```diff
@@ -185,121 +166,189 @@ def __enter__(self):
...
+                    self._result_dict,  # Line result dict of the batch run.
+                    i,  # Index of the sub process.
+                    self._input_queues[i],  # Specific input queue for sub process, used to send input data to it.
+                    self._output_queues[i],  # Specific output queue for sub process, used to receive results from it.
+                ),
+            )
+            for i in range(self._n_process)
+        ]
+
+    def close(self):
```
Comment (Hhhilulu) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:194]: Iterate self._process info. If not empty, force kill the process.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 7 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:267]: The threads will send terminate signal to process after receieving the terminate signal.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 8 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:194]: Added function to process manager.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 9]
```diff
@@ -185,121 +166,192 @@ def __enter__(self):
...
-                if remaining_execution_time <= 0:
-                    self._is_timeout = True
+            while True:
+                try:
+                    # Get task from task_queue
+                    data = task_queue.get(timeout=1)
+                    # Calculate the line timeout for the current line.
+                    line_timeout_sec = self._calculate_line_timeout_sec()
+                    # If the task is a terminate signal or the batch run is timeouted, exit the loop.
+                    if data == TERMINATE_SIGNAL or line_timeout_sec is None:
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:310]: We may extract this part to something like `terminate_process`?

**Review by thy09 — APPROVED** [review]

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 10]
```diff
@@ -316,22 +368,18 @@ def _monitor_workers_and_process_tasks_in_thread(
                 if crashed:
                     bulk_logger.warning(f"Process crashed while executing line {line_number}.")
                     ex = ProcessCrashError(line_number)
-                else:
+                elif self._line_timeout_expired(start_time, line_timeout_sec=line_timeout_sec):
                     # Handle line execution timeout.
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:372]: We may extract these part to something like `handle_output`

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 11]
```diff
@@ -185,121 +166,192 @@ def __enter__(self):
...
-
-            # Put task into input_queue
-            args = (inputs, line_number, run_id, line_timeout_sec)
-            input_queue.put(args)
+                except queue.Empty:
+                    pass
 
-            self._processing_idx[line_number] = format_current_process_info(process_name, process_id, line_number)
-            log_process_status(process_name, process_id, line_number)
+            if exit_loop:
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:334]: Maybe "terminated"?

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py [hunk 12]
```diff
@@ -185,121 +166,192 @@ def __enter__(self):
...
+            self._processing_idx[line_number] = format_current_process_info(process_name, process_id, line_number)
+            log_process_status(process_name, process_id, line_number)
             # Responsible for checking the output queue messages and processing them within a specified timeout period.
-            while not self._batch_timeout_expired(batch_start_time) and not self._line_timeout_expired(start_time):
+            while not self._line_timeout_expired(start_time, line_timeout_sec=line_timeout_sec):
                 # Monitor process aliveness.
-                crashed = not self._is_process_alive(process_id)
+                crashed = not self._processes_manager.is_process_alive(process_id)
                 if crashed:
                     break
```
Comment (thy09) [inline src/promptflow/promptflow/executor/_line_execution_process_pool.py:349]: Overall we may try to make the whole function more clean, to show the core logic, maybe something like this:
```
while not terminated:
    task = get_task()
    if terminated:
        terminate_process()
    put_task_to_input_queue()
    output = get_result_from_output_queue(timeout=xxx)
    handle_output(output)
```

Comment (thy09) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:349]: Currently the code in this function is not that clear to show the main loop.

**Review by Hhhilulu — APPROVED** [review]

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 1 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:310]: Added.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 11 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Updated.

**Review by thy09 — APPROVED** [review]

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 10 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:366]: Will refine in teh next PR.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 9 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py]: Will refine in teh next PR.

File: src/promptflow/promptflow/executor/_line_execution_process_pool.py (hunk 12 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/executor/_line_execution_process_pool.py:343]: Will refine in teh next PR.

**Review by Hhhilulu — APPROVED** [review]

**Pull request merged**